crates/flipperzero/src/storage.rs

```diff
@@ -0,0 +1,252 @@
+        // operations, so we have some logic here to drop any extraneous
+        // information. The possible open modes form a partial order (for
+        // example, `create_new` is more specialized than `truncate`) so we
+        // search for the first "on" bit in this sequence, and use that as the
+        // open mode.
+        let canonicalized_open_mode = if self.open_mode & sys::FS_OpenMode_FSOM_CREATE_NEW != 0 {
+            sys::FS_OpenMode_FSOM_CREATE_NEW
+        } else if self.open_mode & sys::FS_OpenMode_FSOM_CREATE_ALWAYS != 0 {
+            sys::FS_OpenMode_FSOM_CREATE_ALWAYS
+        } else if self.open_mode & sys::FS_OpenMode_FSOM_OPEN_APPEND != 0 {
+            sys::FS_OpenMode_FSOM_OPEN_APPEND
+        } else if self.open_mode & sys::FS_OpenMode_FSOM_OPEN_ALWAYS != 0 {
+            sys::FS_OpenMode_FSOM_OPEN_ALWAYS
+        } else {
+            sys::FS_OpenMode_FSOM_OPEN_EXISTING
+        };
+
+        let f = File::new();
+        if unsafe {
+            sys::storage_file_open(
+                f.0,
+                path.as_ptr() as *const i8,
+                self.access_mode,
+                canonicalized_open_mode,
+            )
+        } {
+            Ok(f)
+        } else {
+            // Per docs, "you need to close the file even if the open operation
+            // failed," but this is handled by `Drop`.
+            Err(Error::from_sys(unsafe { sys::storage_file_get_error(f.0) }))
+        }
+    }
+}
+
+/// File stream with buffered read operations.
+pub struct File(*mut sys::File);
+
+impl File {
+    pub fn new() -> Self {
+        unsafe {
+            File(sys::storage_file_alloc(
+                UnsafeRecord::open(RECORD_STORAGE).as_ptr(),
+            ))
+        }
+    }
+}
+
+impl Drop for File {
+    fn drop(&mut self) {
+        unsafe {
+            // `storage_file_close` calls `storage_file_sync`
+            // internally, so it's not necesssary to call it here.
+            sys::storage_file_close(self.0);
+        }
+    }
+}
+
+impl Read for File {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        let to_read = buf.len().try_into().map_err(|_| Error::InvalidParameter)?;
+        let bytes_read =
+            unsafe { sys::storage_file_read(self.0, buf.as_mut_ptr() as *mut c_void, to_read) };
+        if bytes_read == to_read || bytes_read < to_read && unsafe { sys::storage_file_eof(self.0) }
+        {
+            Ok(bytes_read as usize)
+        } else {
+            Err(Error::from_sys(unsafe {
+                sys::storage_file_get_error(self.0)
+            }))
+        }
+    }
+}
+
+impl Seek for File {
+    fn seek(&mut self, pos: SeekFrom) -> Result<usize, Error> {
+        let (offset_type, offset) = match pos {
+            SeekFrom::Start(n) => (true, n.try_into().map_err(|_| Error::InvalidParameter)?),
+            SeekFrom::End(n) => (false, n.try_into().map_err(|_| Error::InvalidParameter)?),
```

Comment:
Setting `offset_type = false` here will cause `SeekFrom::End` to be treated the same way as `SeekFrom::Current` (because `offset_type` is really `from_start`).

Instead, for `SeekFrom::End` we will need to measure the length of the file, and then use `from_start = true` and `offset = file_length - n`.

Reply:
I'm realizing that this is all subtly incorrect if we're going off of the Rust semantics. In particular, the user should be able to specify a negative number for `End` or `Current`, but the current code doesn't allow that.

My problem is that the `offset` parameter in `storage_file_seek` is a `u32` whereas the rest of the API treats the file offset as a `u64`, so it isn't super clear how to calculate `offset = file_length - n` such that it can be passed to `storage_file_seek`.

I'll see if there are any examples upstream

Reply:
I couldn't find many, but there's at least one example of necking `storage_file_size` down to a `uint32_t` in `lib/toolbox/crc32_calc.c`.

The LittleFS 'file max' is a 32-bit int and `lfs_file_size` returns a `int32_t`, so I'm tempted to say that we can assume `storage_file_size` and friends return a value in `[0, u32::MAX]`. Does that sound right? Or is it bad to assume that the firmware is always going to be on LittleFS